### ISSUE TITLE
Add generate_typelib in the GNOME module

### ISSUE THREAD

**Issue by arteymix** [mesonbuild/meson#1014]
This would cover Vala and any tool that produce custom GIR.

We should deprecate `generate_gir` and split it into two distinct tools. This will have a few advantages like not returning two targets or passing redundant arguments.

**Comment by ptomato** [comment]
This would also be useful for projects that generate private typelibs and don't want to install the GIR because it's not public API.

You can kind of fake this with `install_dir_gir: '/dev/null', install_dir_typelib: typelibdir`, but that breaks `DESTDIR` so it's not a really workable solution.

**Comment by ptomato** [comment]
It seems that `install_dir_gir: false` works for that purpose, no idea if that's intentional / liable to break or not.

**Comment by arteymix** [comment]
I believe it's intended. We have the same behaviour for Vala `install_dir` with a special `true`-implies-default-location behaviour.

**Comment by esodan** [comment]
But the issue here is to generate the typelib. It is generated by a different command and depends on the library compilation.

**Comment by albfan** [comment]
Based on existing use case for vala:

```meson
foo_lib = shared_library('foo', 'foo.vala',
                  vala_gir: 'Foo-1.0.gir',
                  dependencies: [glib_dep, gobject_dep],
                  install: true,
                  install_dir: [true, true, true, true])

g_ir_compiler = find_program('g-ir-compiler')
typelib_input = meson.current_build_dir() / 'Foo-1.0.gir'
custom_target('foo typelib', command: [g_ir_compiler, '--output', '@OUTPUT@', typelib_input],
              output: 'Foo-1.0.typelib',
              depends: foo_lib,
              install: true,
              install_dir: get_option('libdir') / 'girepository-1.0')
```

shared_library is a general object, but it knows about `vala_gir`, as vala needs to create gir file with its own compiler

Can `gnome.generate_gir` know about a shared_library result is built with vala, so it uses its gir file and only creates the typelib part?

This back to standard how people use `gnome.generate_gir` and avoid developer to point to generated files in build directory.

If that's not possible, this function might be:

```meson
foo_lib = shared_library('foo', 'foo.vala',
                  vala_gir: 'Foo-1.0.gir',
                  dependencies: [glib_dep, gobject_dep],
                  install: true,
                  install_dir: [true, true, true, true]
)

gnome = import('gnome')

gnome.generate_typelib(foo_lib, 
  gir:  meson.current_build_dir() / 'Foo-1.0.gir',
  namespace: 'Foo', #this can be read from gir?
  nsversion: '1.0', #this can be read from gir?
  depends: foo_lib,
  install: true,
  install_dir: get_option('libdir') / 'girepository-1.0'
)
```

The more info we can infer from lib result, the better to avoid this annoyance

**Comment by dcbaker** [comment]
@albfan yes, it could. See https://github.com/mesonbuild/meson/pull/14782, which would simplify this case for you already

**Comment by albfan** [comment]
Nice, if #14782 is in place, I think gnome `generate_gir` can then adapt and generate only typelib target.

Don't know other use cases @ptomato mentions that might make `generate_typelib` interesting.

I will track this solution in https://github.com/albfan/vala-libpeas-demo, improving as things get merged, and trying to push on gnome documentation

**Comment by dcbaker** [comment]
@albfan @ptomato  I wrote up https://github.com/dcbaker/meson/tree/submit/gnome-generate-typelib, after looking at `gnome.generate_gir`. I kinda like the idea of just a "here is a GIR file, create the matching typelib file for me".

One of the reasons for that is that there's a ton of information that is passed to the generate_gir command that we can extract directly from the GIR file. There's definately a few things left we could do better, but internally it does a bunch of code sharing with the `generate_gir` method.

I'm still not 100% sure what the most ergonomic way to halde it is.

**Comment by ptomato** [comment]
My use case is pretty much covered by `install_dir_gir: false`.
